mla/threeml/sob_terms.py

```diff
@@ -0,0 +1,193 @@
+
+    @property
+    def sob(self) -> np.ndarray:
+        """Docstring"""
+        return self._energysobhist[self._sin_dec_idx, self._log_energy_idx]
+
+
+@dataclasses.dataclass
+class ThreeMLPSEnergyTermFactory(sob_terms.SoBTermFactory):
+    """Docstring"""
+    data_handler: data_handlers.ThreeMLDataHandler
+    source: sources.PointSource
+    _spectrum: spectral.BaseSpectrum = dataclasses.field(init=False, repr=False)
+    _bg_sob: np.ndarray = dataclasses.field(init=False, repr=False)
+    _reduced_reco_sim: np.ndarray = dataclasses.field(init=False, repr=False)
+    _sin_dec_bins: np.ndarray = dataclasses.field(init=False, repr=False)
+    _log_energy_bins: np.ndarray = dataclasses.field(init=False, repr=False)
+    _bins: np.ndarray = dataclasses.field(init=False, repr=False)
+
+    def __post_init__(self) -> None:
+        """Docstring"""
+        self._init_bg_sob_map()
+        self.spectrum = self.config['assumed_spectrum']
+        self._sin_dec_bins = np.linspace(-1, 1, 1 + self.config['sin_dec_bins'])
+        self._log_energy_bins = np.linspace(
+            *self.config['log_energy_bounds'], 1 + self.config['log_energy_bins'])
+        self._reduced_reco_sim = self.data_handler.cut_reconstructed_sim(
+            self.source.location[1], *self.config['reco_sampling_width'])
+        self._bins = np.array([self._sin_dec_bins, self._log_energy_bins])
+
+    def __call__(
+        self,
+        params: par.Params,
+        events: np.ndarray
+    ) -> sob_terms.SoBTerm:
+        """Docstring"""
+        sin_dec_idx = np.searchsorted(self._sin_dec_bins[:-1], events['sindec'])
+        log_energy_idx = np.searchsorted(self._log_energy_bins[:-1], events['logE'])
+
+        # Get the bin that each event belongs to
+        try:
+            sin_dec_idx = np.searchsorted(self._sin_dec_bins[:-1],
+                                          events['sindec']) - 1
+        except ValueError:
+            sin_dec_idx = np.searchsorted(self._sin_dec_bins[:-1],
+                                          np.sin(events['dec'])) - 1
+
+        log_energy_idx = np.searchsorted(self._log_energy_bins[:-1],
+                                         events['logE']) - 1
+
+        return ThreeMLPSEnergyTerm(
+            _params=params,
+            _sob=np.empty(1),
+            _sin_dec_idx=sin_dec_idx,
+            _log_energy_idx=log_energy_idx,
+            _energysobhist=self.cal_sob_map(),
+        )
+
+    def _init_bg_sob_map(self) -> np.ndarray:
+        """Docstring"""
+        bg_h = self.data_handler.build_background_sindec_logenergy_histogram(
+            self._bins
+        )
+        # Normalize histogram by dec band
+        bg_h /= np.sum(bg_h, axis=1)[:, None]
+        self._bg_sob = bg_h
+
+    def update_source(
+        self,
+        source: sources.PointSource
+    ) -> None:
+        """Docstring"""
+        self.source = source
+        self._reduced_reco_sim = self.data_handler.cut_reconstructed_sim(

```

Comment:
this should be stored in the datahandler. we should keep the actual structure of the data abstracted away

Reply:
Here the reduced_reco_sim is only used when evaluating the energy signal histogram and would never use in the injection. Also, it will change when the source hypothesis changed.  The analog of this _reduced_reco_sim should be _spline_map in SplineMapEnergyTermFactory and I think we should put this term in the SOB factory